File: packages/api/op_agent/geo.py
"""縣市／行政區解析。

把使用者的口語地址轉成統一資訊資料集裡的 county_code / district_code，
這樣服務單、廠商服務區、諮詢單就能對得起來。
"""

from __future__ import annotations

from .domain import Address
from .geo_generated import COUNTIES, DISTRICTS

_COUNTY_BY_CODE = {c["code"]: c for c in COUNTIES}
_DISTRICT_BY_CODE = {d["code"]: d for d in DISTRICTS}


def _normalize(name: str) -> str:
    """台北/臺北、台中/臺中 等異體字正規化。"""
    return name.strip().replace("臺", "台").replace(" ", "")


def find_county(text: str) -> dict | None:
    q = _normalize(text)
    if not q:
        return None
    for c in COUNTIES:
        if _normalize(c["name"]) == q:
            return {"code": c["code"], "name": c["name"]}
    # 「台北」→「台北市」
    for c in COUNTIES:
        n = _normalize(c["name"])
        if n.startswith(q) or q.startswith(n):
            return {"code": c["code"], "name": c["name"]}
    return None


def find_district(county_code: str, text: str) -> dict | None:
    q = _normalize(text)
    if not q:
        return None
    pool = [d for d in DISTRICTS if d["county_code"] == county_code]
    for d in pool:
        if _normalize(d["name"]) == q:
            return {"code": d["code"], "name": d["name"]}
    for d in pool:
        n = _normalize(d["name"])
        if n.startswith(q) or q.startswith(n):
            return {"code": d["code"], "name": d["name"]}
    return None


def list_districts(county_code: str) -> list[str]:
    return [d["name"] for d in DISTRICTS if d["county_code"] == county_code]


def resolve_address(
    *,
    county: str | None = None,
    district: str | None = None,
    detail: str | None = None,
    free_text: str | None = None,
) -> Address | None:
    """把使用者口語地址解析成 Address（含 county/district code）。

    例：「台北市大安區復興南路一段100號5樓」
    """
    county_raw = county
    district_raw = district

    if free_text and (not county_raw or not district_raw):
        text = _normalize(free_text)
        hit = next((c for c in COUNTIES if _normalize(c["name"]) in text), None)
        if hit:
            county_raw = hit["name"]
            idx = text.index(_normalize(hit["name"])) + len(_normalize(hit["name"]))
            rest = text[idx:]
            d_hit = next(
                (
                    d
                    for d in DISTRICTS
                    if d["county_code"] == hit["code"] and rest.startswith(_normalize(d["name"]))
                ),
                None,
            )
            if d_hit:
                district_raw = d_hit["name"]
                if detail is None:
                    detail = rest[len(_normalize(d_hit["name"])) :]
            elif detail is None:
                detail = rest

    if not county_raw:
        return None
    c = find_county(county_raw)
    if not c:
        return None
    d = find_district(c["code"], district_raw) if district_raw else None
    if not d:
        return None

    addr: Address = {
        "countyCode": c["code"],
        "countyName": c["name"],
        "districtCode": d["code"],
        "districtName": d["name"],
    }
    cleaned = (detail or "").strip()
    if cleaned:
        addr["detail"] = cleaned
    return addr


def describe_address(a: Address) -> str:
    return f"{a.get('countyName', '')}{a.get('districtName', '')}{a.get('detail', '')}"


def county_name(code: str) -> str:
    row = _COUNTY_BY_CODE.get(code)
    return row["name"] if row else code


def district_name(code: str) -> str:
    row = _DISTRICT_BY_CODE.get(code)
    return row["name"] if row else code

File: packages/backend/src/data/geo.ts
import type { Address } from '../domain/types';
import { COUNTIES, DISTRICTS } from './geo.generated';

/** 台北/臺北、台中/臺中 等異體字正規化 */
function normalize(name: string): string {
  return name.trim().replace(/臺/g, '台').replace(/\s/g, '');
}

const countyByCode = new Map(COUNTIES.map((c) => [c.code, c]));
const districtByCode = new Map(DISTRICTS.map((d) => [d.code, d]));

export function findCounty(input: string): { code: string; name: string } | undefined {
  const q = normalize(input);
  const exact = COUNTIES.find((c) => normalize(c.name) === q);
  if (exact) return exact;
  // 「台北」→「台北市」
  return COUNTIES.find((c) => normalize(c.name).startsWith(q) || q.startsWith(normalize(c.name)));
}

export function findDistrict(
  countyCode: string,
  input: string,
): { code: string; name: string } | undefined {
  const q = normalize(input);
  const pool = DISTRICTS.filter((d) => d.countyCode === countyCode);
  const exact = pool.find((d) => normalize(d.name) === q);
  if (exact) return exact;
  return pool.find((d) => normalize(d.name).startsWith(q) || q.startsWith(normalize(d.name)));
}

export function listDistricts(countyCode: string): string[] {
  return DISTRICTS.filter((d) => d.countyCode === countyCode).map((d) => d.name);
}

/**
 * 把使用者口語地址解析成 Address（含統一資訊的 county/district code）。
 * 例：「台北市大同區承德路三段10號」
 */
export function resolveAddress(input: {
  county?: string;
  district?: string;
  detail?: string;
  freeText?: string;
}): Address | undefined {
  let countyRaw = input.county;
  let districtRaw = input.district;
  let detail = input.detail;

  if (input.freeText && (!countyRaw || !districtRaw)) {
    const text = normalize(input.freeText);
    const county = COUNTIES.find((c) => text.includes(normalize(c.name)));
    if (county) {
      countyRaw = county.name;
      const rest = text.slice(text.indexOf(normalize(county.name)) + county.name.length);
      const district = DISTRICTS.filter((d) => d.countyCode === county.code).find((d) =>
        rest.startsWith(normalize(d.name)),
      );
      if (district) {
        districtRaw = district.name;
        detail = detail ?? rest.slice(district.name.length);
      } else {
        detail = detail ?? rest;
      }
    }
  }

  if (!countyRaw) return undefined;
  const county = findCounty(countyRaw);
  if (!county) return undefined;

  const district = districtRaw ? findDistrict(county.code, districtRaw) : undefined;
  if (!district) return undefined;

  return {
    countyCode: county.code,
    countyName: county.name,
    districtCode: district.code,
    districtName: district.name,
    detail: detail?.trim() || undefined,
  };
}

export function describeAddress(a: Address): string {
  return `${a.countyName}${a.districtName}${a.detail ?? ''}`;
}

export function countyName(code: string): string {
  return countyByCode.get(code)?.name ?? code;
}

export function districtName(code: string): string {
  return districtByCode.get(code)?.name ?? code;
}
